alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,180 @@
+
+use std::io::Error;
+use std::os::raw::c_void;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+use log::warn;
+
+use mio_extras::channel::{channel, Receiver};
+
+use winapi::shared::ntdef::{BOOLEAN, HANDLE, PVOID};
+use winapi::um::winbase::{RegisterWaitForSingleObject, UnregisterWait, INFINITE};
+use winapi::um::winnt::{WT_EXECUTEINWAITTHREAD, WT_EXECUTEONLYONCE};
+
+use crate::tty::ChildEvent;
+
+/// WinAPI callback for `HandleWaitSignal`, unpacks Rust closure reference and calls it
+extern "system" fn child_exit_callback<F>(ctx: PVOID, timed_out: BOOLEAN)
+where
+    F: FnOnce() + Send,
+{
+    if timed_out != 0 {
+        return;
+    }
+
+    let callback: Box<F> = unsafe { Box::from_raw(ctx as *mut F) };
+    callback();
+}
+
+/// Represents closure attached to Win32 handle wait signal
+///
+/// This allows to fire a Rust callback when subprocess exits
+pub(crate) struct HandleWaitSignal {
+    wait_handle: AtomicPtr<c_void>,
+}
+
+impl HandleWaitSignal {
+    /// Registers an asynchronous closure to call when `child_handle` wait signals.
+    ///
+    /// The `on_exit` is called on Win32 threadpool thread so it should avoid
+    /// blocking calls. See [`WT_EXECUTEINWAITTHREAD` flag docs] for details
+    ///
+    /// [`WT_EXECUTEINWAITTHREAD` flag docs]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerwaitforsingleobject
+    fn new<F>(child_handle: HANDLE, on_exit: F) -> Result<HandleWaitSignal, Error>
+    where
+        F: FnOnce() + Send,
+    {
+        let mut wait_handle: HANDLE = 0 as HANDLE;
+        let wait_notify = Box::new(on_exit);
+
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback::<F>),
+                Box::into_raw(wait_notify) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+pub(crate) struct ChildProcessState {
+    _on_exit: HandleWaitSignal,
+    events: Receiver<ChildEvent>,
+}
+
+impl ChildProcessState {
+    pub fn new(subprocess_handle: HANDLE) -> Result<ChildProcessState, Error> {
+        let (sender, receiver) = channel();
+        let on_exit = HandleWaitSignal::new(subprocess_handle, move || {
+            if let Err(e) = sender.send(ChildEvent::Exited) {
+                warn!(
+                    "An error occurred while attempting to notify about child process z
+                    termination: {}",
+                    e
+                );
```

Comment:
When would this happen? I think the only case a `send` can fail is when the receiver is dead. Do we need to handle that?

What's the argument against using `let _ = sender.send(ChildEvent::Exited)`? Do we require this to be sent successfully? I'd assume if the other end isn't receiving anymore, it has already exited anyways.

Reply:
This will probably only fail if the Alacritty process is already exiting and due to some unlucky context switching the receiving end gets cleaned up before callback is deregistered or the deregistering fails. It's probably not worth panicking the thread over (esp. that this is executed in WinAPI thread pool thread), but I thought it might be worth logging. Could be just ignored if you think that's a better solution.

Reply:
> but I thought it might be worth logging

Why? For issues potentially introduced in the future? As far as I can tell this should never be something we care about in the present implementation, which is why it's probably cleaner to just swallow it.

Reply:
Removed in d1a2dfc